Plutarch/Field/HList.hs

```diff
@@ -0,0 +1,193 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Plutarch.Field.HList (
+  -- * HList and HRec types
+  HList (..),
+  HRec (..),
+
+  -- * Field indexing functions
+  hlistField,
+  hrecField,
+
+  -- * Type families
+  type IndexList,
+  type IndexOf,
+  type SingleItem,
+  type Take,
+  type Drop,
+  type Range,
+
+  -- * Internal utils
+  Elem (..),
+  indexHList,
+  NatElem (..),
+) where
+
+import Data.Kind (Type)
+import GHC.Records (HasField (..))
+import GHC.TypeLits (
+  Nat,
+  Symbol,
+  type (+),
+  type (-),
+ )
+
+--------------------------------------------------------------------------------
+---------- HList and HRec types
+
+-- | Usual GADT Heterogenous List encoding
+data HList (xs :: [Type]) where
+  HNil :: HList '[]
+  HCons :: x -> HList xs -> HList (x ': xs)
+
+{- |
+  Heterogenous Record, using a list of symbols as a
+  set of corresponding indices by position.
+-}
+newtype HRec (fs :: [Symbol]) (as :: [Type])
```

Comment:
i think it would be simpler if you merged `fs` and `as` like done elsewhere. You can remove `HList` too.

Reply:
Went back on forth on how relevant the internal HList would be... it would take a little rewrite to have the more usual encoding... probably worthwhile though